lib/execution_engine2/sdk/EE2Runjob.py

```diff
@@ -258,14 +248,150 @@ def _prepare_to_run(self, params, concierge_params=None) -> JobSubmissionParamet
             source_ws_objects=params.get(_SOURCE_WS_OBJECTS),
         )
 
+    def _prepare_to_run(self, params, concierge_params=None) -> JobSubmissionParameters:
+        """
+        Creates a job record and creates the job submission params
+        """
+
+        job_id = self._init_job_rec(self.sdkmr.get_user_id(), params)
+        self.logger.debug(
+            f"User {self.sdkmr.get_user_id()} attempting to run job {params[_METHOD]} {params}"
+        )
+        return self._generate_job_submission_params(job_id, params)
+
+    def _run_multiple(self, runjob_params):
+        """
+        Get the job records, bulk save them, then submit to condor.
+        If any condor submission fails, abort all of the jobs
+        :return:
+        """
+        # Save records to db
+        init_job_rec = time.time()
+        job_records = []
+        for runjob_param in runjob_params:
+            job_records.append(
+                self._init_job_rec(self.sdkmr.get_user_id(), runjob_param, save=False)
+            )
+        self.logger.debug(f"init_job_rec = {time.time() - init_job_rec}")
+
+        save_jobs = time.time()
+        job_ids = self.sdkmr.save_jobs(job_records)
+
+        self.logger.debug(f"save_jobs = {time.time() - save_jobs}")
+        self.logger.debug(f"init and save save_jobs = {time.time() - init_job_rec}")
+
+        # Generate job submission params
+        gen_sub_time = time.time()
+        job_submission_params = []
+        for i, job_id in enumerate(job_ids):
+            job_submission_params.append(
+                self._generate_job_submission_params(job_id, runjob_params[i])
+            )
+            assert job_id == job_submission_params[i].job_id
+
+        self.logger.debug(
+            f"init and gen_sub_time save_jobs = {time.time() - gen_sub_time}"
+        )
+
+        # Takes 2.5200018882751465 for 100, can shave off 2.5 secs by making this async
+        kafku = time.time()
+        for job_id in job_ids:
+            self.sdkmr.get_kafka_client().send_kafka_message(
+                message=KafkaCreateJob(
+                    job_id=str(job_id), user=self.sdkmr.get_user_id()
+                )
+            )
+
+        self.logger.debug(f"kafka submit = {time.time() - kafku}")
+
+        # Submit to Condor
+        condor_time = time.time()
+        try:
+            submission_ids = self._submit_multiple(job_submission_params)
+            self.logger.debug(f"condor_time submit = {time.time() - condor_time}")
+            return submission_ids
+        except Exception as e:
+            self._abort_multiple_jobs(job_ids)
+            raise e
+
+    def _update_to_queued_multiple(self, job_ids, scheduler_ids):
+        # TODO Unused
+        if len(job_ids) != len(scheduler_ids):
+            raise Exception(
+                "Need to provide the same amount of job ids and scheduler_ids"
+            )
+
+        # TODO RETRY FOR RACE CONDITION OF RUN/CANCEL
```

Comment:
Can you explain this? This sounds like it could cause problems

Reply:
On second thought I don't think there can be a race condition, because you don't know the job ids to cancel until the method returns.

Reply:
But if you thread everything after the initial mongo insert then there could be, right?

Reply:
Right.. So if while you are running these jobs, you hit cancel, then there's a thread that's still going and updating the jobs to "queued".. So cancel won't actually cancel them all, and it might not be able to kill the jobs from condor because it hasn't been submitted to condor yet.

Reply:
Right, so even if we thread the updating code as it stands now, there is still a small window in which it's possible to call update on a job that should be cancelled. It would probably be an even smaller window than this pr.

Reply:
Hmm, so for a lot of jobs where the thread is running for a while, a user could cancel a job before the thread has set the update state, and then the thread would come along and blow away the canceled state.
Are there any other fields other than the state that could be wrong? Like a finished timestamp?

Right now I'm thinking that this is out of scope for this PR and we should just make an issue and deal with it in a later PR

Reply:
It shouldn't blow away the canceled state, since the Update statement can't work, since it updates only records that are in the state of "Created", and if its in the state of cancelled, it can't update.

Reply:
Right, I wrote that before I looked at the most recent diffs

Reply:
So is this issue fixed? It seems like if a race condition does happen and a job is cancelled before it's updated the mongo record stays as cancelled

Reply:
Oh but the condor job might keep running

Reply:
* If I am understanding the way mongo updateone statements believe, I believe so.
*  It's not possible to cancel the condor job if we don't know its scheduler_id.. It will silently fail and send some poorly formed kafka messages
* The condor job has the job cancellation checker though, so it should get cancelled that way if it checks the jobs cancellation status in ee2 db

Reply:
Ok, that last one makes this seem fine. I didn't notice a test showing that jobs not in the created state don't get updated though, maybe add that?

Reply:
Changed the assert and added tests.

Reply:
:+1: